### PR TITLE
feat: gross profit and operating income lines on PL statement, provisional income on balance sheet

### PR DESCRIPTION
**Pull request by rvpasari**
https://app.asana.com/0/1198955099016653/1199000235912998

### CODE REVIEW

**Comment by rvpasari** [comment]
Balance Sheet

File: erpnext/setup/doctype/company/company.json [hunk 1]
```diff
@@ -736,13 +738,27 @@
...
+   "label": "Default Direct Expenses",
+   "options": "Account"
+  },
+  {
+   "depends_on": "eval:!doc.__islocal",
+   "fieldname": "default_indirect_expenses",
+   "fieldtype": "Link",
+   "ignore_user_permissions": 1,
+   "label": "Default Indirect Expenses",
+   "options": "Account"
```
Comment (hrwX) [inline erpnext/setup/doctype/company/company.json:754]: marked this as dependent PR, this change is present in another PR of yours too.

File: erpnext/accounts/report/balance_sheet/balance_sheet.py [hunk 2]
```diff
@@ -31,9 +31,26 @@ def execute(filters=None):
...
+
+	#add provisional profit/loss and adjust equity totals if books are not closed
+	if len(equity) > 2 and provisional_profit_loss:
+		equity = insert_equity_provisions(equity, provisional_profit_loss, period_list)
+		data.extend(equity or [])
+
+	#specific case when equity is completely empty but we still need to add
+	elif not equity and provisional_profit_loss:
+		equity = create_equity_provisions(provisional_profit_loss, period_list)
+		data.extend(equity)
```
Comment (hrwX) [inline erpnext/accounts/report/balance_sheet/balance_sheet.py]: create and insert equity provisions mean the same thing

File: erpnext/accounts/report/balance_sheet/balance_sheet.py [hunk 3]
```diff
@@ -49,17 +66,62 @@ def execute(filters=None):
...
+	equity = [{"account_name": "Equity", "account": "Equity", "indent": 0.0, "is_group": 1.0}]
+	provisional_profit_loss['parent_account']= "Equity"
+	provisional_profit_loss["indent"] = 1.0
+	equity.append(provisional_profit_loss)
+	total_summary = {"account_name": "Total Equity", "account": "Total Equity", "indent": 0.0, "is_group": 0.0}
+	for period in period_list:
+		if period.key in provisional_profit_loss:
+			total_summary[period.key] = provisional_profit_loss[period.key]
+	equity.append(total_summary)
+	equity.append({})
```
Comment (hrwX) [inline erpnext/accounts/report/balance_sheet/balance_sheet.py]:
```suggestion
```

File: erpnext/accounts/report/balance_sheet/balance_sheet.py [hunk 4]
```diff
@@ -68,10 +130,10 @@ def get_provisional_profit_loss(asset, liability, equity, period_list, company,
 		total = total_row_total=0
 		currency = currency or frappe.get_cached_value('Company',  company,  "default_currency")
 		total_row = {
-			"account_name": "'" + _("Total (Credit)") + "'",
-			"account": "'" + _("Total (Credit)") + "'",
+			"account_name": "'" + _("Total Liabilities and Equity") + "'",
+			"account": "'" + _("Total Liabilities and Equity") + "'",
```
Comment (hrwX) [inline erpnext/accounts/report/balance_sheet/balance_sheet.py]:
```suggestion
			"account_name": _("Total Liabilities and Equity"),
			"account": _("Total Liabilities and Equity"),
```

File: erpnext/accounts/report/balance_sheet/balance_sheet.py [hunk 5]
```diff
@@ -68,10 +130,10 @@ def get_provisional_profit_loss(asset, liability, equity, period_list, company,
...
 			"warn_if_negative": True,
-			"currency": currency
+			"currency": currency,
```
Comment (hrwX) [inline erpnext/accounts/report/balance_sheet/balance_sheet.py]:
```suggestion
			"currency": currency
```

File: erpnext/accounts/report/balance_sheet/balance_sheet.py [hunk 6]
```diff
@@ -100,8 +162,13 @@ def get_provisional_profit_loss(asset, liability, equity, period_list, company,
...
 				"warn_if_negative": True,
-				"currency": currency
-			})
+				"currency": currency,
+				"total": None,
+				"is_group": 0.0,
+				"has_value": True,
+				"opening_balance": -0.0,
+				"parent_account": "Equity"
+				})
```
Comment (hrwX) [inline erpnext/accounts/report/balance_sheet/balance_sheet.py]:
```suggestion
			})
```

File: erpnext/accounts/report/financial_statements.py [hunk 7]
```diff
@@ -228,17 +229,20 @@ def get_data(company, root_type, balance_must_be, period_list, filters=None,
...
-	"""				
+	"""
 	#get the list of accounts associated with a particular root (ex. Asset, Liability, etc)
 	accounts = get_accounts(company, root_type)
 	if not accounts:
 		return None
 
+	# if root_type == "Equity":
+	# 	print("Equity accounts are: ", accounts)
+
```
Comment (hrwX) [inline erpnext/accounts/report/financial_statements.py]:
```suggestion
```

File: erpnext/accounts/report/financial_statements.py [hunk 8]
```diff
@@ -480,24 +485,31 @@ def generate_parent_account_totals(out, root_type, balance_must_be, period_list,
 				"is_group":0
 			})
 
+		#print(row.get("account_name"))
+		if row.get("account_name") == "650000 - Other Expenses":
+			print("................................")
+			print("Processing other expenses.....")
+			print(row)
+			print(parent_accounts[-1])
```
Comment (hrwX) [inline erpnext/accounts/report/financial_statements.py]:
```suggestion
```

File: erpnext/accounts/report/financial_statements.py [hunk 9]
```diff
@@ -480,24 +485,31 @@ def generate_parent_account_totals(out, root_type, balance_must_be, period_list,
 				"is_group":0
 			})
 
+		#print(row.get("account_name"))
```
Comment (hrwX) [inline erpnext/accounts/report/financial_statements.py]:
```suggestion
```

File: erpnext/accounts/report/profit_and_loss_statement/profit_and_loss_statement.py [hunk 10]
```diff
@@ -23,20 +25,91 @@ def execute(filters=None):
...
+	#get direct and indirect expense account names
+	gross_profit_loss = None
+	direct_expense_account = frappe.get_value("Company", get_default_company(), "default_direct_expenses")
+	indirect_expense_account = frappe.get_value("Company", get_default_company(), "default_indirect_expenses")
+	direct_expenses = None
+	for expense_item in expense:
+		data.append(expense_item)
+
+		#add gross profit line item on the PL statement
+		if expense_item.get("account_name") == "Total  "+direct_expense_account.split("-")[1].strip():
```
Comment (hrwX) [inline erpnext/accounts/report/profit_and_loss_statement/profit_and_loss_statement.py]: - Is there no better way to do this ?

**Review by hrwX — CHANGES_REQUESTED** [review]

File: erpnext/setup/doctype/company/company.json (hunk 1 above)
Comment (rvpasari) [reply erpnext/setup/doctype/company/company.json:754]: yes, thanks for noticing. It slipped my attention otherwise I usually mark this as dependent

File: erpnext/accounts/report/profit_and_loss_statement/profit_and_loss_statement.py (hunk 10 above)
Comment (rvpasari) [reply erpnext/accounts/report/profit_and_loss_statement/profit_and_loss_statement.py]: I thought about, and the only other place that this change could even go to is the financial_statements.py file. However, since I only have data for Income accounts and Expense Accounts in two different instances, I cannot calculate gross profit / loss and operating profit / loss there. Also need gross profit / loss to calculate operating profit / loss.

File: erpnext/accounts/report/balance_sheet/balance_sheet.py [hunk 11]
```diff
@@ -31,9 +31,26 @@ def execute(filters=None):
...
-	data.extend(equity or [])
+
+	#add provisional profit/loss and adjust equity totals if books are not closed
+	if len(equity) > 2 and provisional_profit_loss:
+		equity = insert_equity_provisions(equity, provisional_profit_loss, period_list)
+		data.extend(equity or [])
+
+	#specific case when equity is completely empty but we still need to add
+	elif not equity and provisional_profit_loss:
+		equity = create_equity_provisions(provisional_profit_loss, period_list)
```
Comment (hrwX) [inline erpnext/accounts/report/balance_sheet/balance_sheet.py]: - Resolve the naming issue.
- Insert and create means the same as it implies that you'll end up inserting data in to db by creating it

**Review by hrwX — CHANGES_REQUESTED** [review]

File: erpnext/accounts/report/balance_sheet/balance_sheet.py [hunk 12]
```diff
@@ -49,17 +66,62 @@ def execute(filters=None):
...
 
 	columns = get_columns(filters.periodicity, period_list, filters.accumulated_values, company=filters.company)
-
 	chart = get_chart_data(filters, columns, asset, liability, equity)
 
 	return columns, data, message, chart
 
+def insert_equity_provisions(equity, provisional_profit_loss, period_list):
+	"""
+	[summary]
```
Comment (sahil28297) [inline erpnext/accounts/report/balance_sheet/balance_sheet.py]: Add summary line

File: erpnext/accounts/report/balance_sheet/balance_sheet.py [hunk 13]
```diff
@@ -49,17 +66,62 @@ def execute(filters=None):
...
+	equity.pop()
+	for period in period_list:
+		if period.key in total_equity and period.key in provisional_profit_loss:
+			total_equity[period.key] = 0.0 if not total_equity[period.key] else total_equity[period.key]
+			total_equity[period.key] += provisional_profit_loss[period.key] #update equity account total for all periods
+	return equity
+
+
+
+def create_equity_provisions(provisional_profit_loss, period_list):
```
Comment (sahil28297) [inline erpnext/accounts/report/balance_sheet/balance_sheet.py]: `insert_equity_provisions` and `create_equity_provisions` sounds almost similar, can you use some better naming for these functions?

File: erpnext/accounts/report/balance_sheet/balance_sheet.py (hunk 3 above)
Comment (sahil28297) [inline erpnext/accounts/report/balance_sheet/balance_sheet.py]: please add a comment as to why you are appending a blank line, people won't understand it at the first glance

File: erpnext/accounts/report/financial_statements.py [hunk 14]
```diff
@@ -18,7 +18,8 @@
 from six import itervalues
 from erpnext.accounts.doctype.accounting_dimension.accounting_dimension import get_accounting_dimensions, get_dimension_with_children
 
-def get_period_list(from_custom_date, to_custom_date, periodicity, accumulated_values=False, 
+non_zero_parent_accounts = {}
```
Comment (sahil28297) [inline erpnext/accounts/report/financial_statements.py]: Is there a need for a global variable, it is used at just one place?

**Review by sahil28297 — DISMISSED** [review]

File: erpnext/accounts/report/profit_and_loss_statement/profit_and_loss_statement.py (hunk 10 above)
Comment (hrwX) [reply erpnext/accounts/report/profit_and_loss_statement/profit_and_loss_statement.py]: There has to be a better way of doing this!
Add a key in expense dict for comparing accounts rather than this

**Review by hrwX — CHANGES_REQUESTED** [review]

File: erpnext/accounts/report/financial_statements.py (hunk 14 above)
Comment (rvpasari) [reply erpnext/accounts/report/financial_statements.py]: No need for this at all, I had added it to test someting, don't need it. Forgot to remove it I guess.

File: erpnext/accounts/report/balance_sheet/balance_sheet.py (hunk 3 above)
Comment (rvpasari) [reply erpnext/accounts/report/balance_sheet/balance_sheet.py]: also done.

File: erpnext/accounts/report/balance_sheet/balance_sheet.py (hunk 13 above)
Comment (rvpasari) [reply erpnext/accounts/report/balance_sheet/balance_sheet.py]: done.

File: erpnext/accounts/report/balance_sheet/balance_sheet.py (hunk 11 above)
Comment (rvpasari) [reply erpnext/accounts/report/balance_sheet/balance_sheet.py]: resolved.

**Comment by neilLasrado** [comment]
@rvpasari you have a Batch_no commit on this PR, is that by mistake? You might have to rebase your branch

**Comment by rvpasari** [comment]
> @rvpasari you have a Batch_no commit on this PR, is that by mistake? You might have to rebase your branch

Can't seen any changes made to anything related to batch_no over here, could you point out the file/line?

**Review by hrwX — APPROVED** [review]

**Review by sahil28297 — APPROVED** [review]
